python3Packages.bitsandbytes: 0.47.0 -> 0.48.1, support rocm

06kellyjac opened the pull request:
Added support for ROCm since the 0.47.0 update https://github.com/bitsandbytes-foundation/bitsandbytes/pull/1683

Made some small refactors to make more consistent with other packages following off `torch.cuda*`

All the ROCm stuff is in one drv, I'm not sure I recognise the need for the split between cuda-common-redist, cuda-native-redist, and cuda-redist.

Maybe needs an assert that cudaSupport and rocmSupport are exclusive? :shrug:

For testing:
```nix
{
  pkgs ? import ./. { config = { rocmSupport = true; }; },
}:
pkgs.python3Packages.bitsandbytes
```

Ran a package that wanted bitsandbytes with ROCm locally (invokeai, still testing other stuff) and at least doesn't complain on launch anymore.

cc: @bcdarwin

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [X] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [X] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

LunNova commented:
> Maybe needs an assert that cudaSupport and rocmSupport are exclusive? 🤷

Could copy what torch does to mark broken. Not sure it's really needed since this depends on torch and in most cases you'll see the error that torch can't build for CUDA and ROCm at the same time.

06kellyjac committed:
python3Packages.bitsandbytes: 0.47.0 -> 0.48.1

Diff: https://github.com/bitsandbytes-foundation/bitsandbytes/compare/0.47.0...0.48.1

Changelog: https://github.com/bitsandbytes-foundation/bitsandbytes/releases/tag/0.48.1

06kellyjac committed:
python3Packages.bitsandbytes: support rocm

06kellyjac committed:
python3Packages.bitsandbytes: add jk to maintainers

SuperSandro2000 reviewed (approved):
LGTM but didn't test
